docs/dev/background-information/sql/upgrades_and_migrator_design.mdx

```diff
@@ -0,0 +1,357 @@
+alleviating a common issue in which frequent transaction locks caused failed migration on Sourcegraph startup. 
+You can learn more in our [migrations overview docs](./migrations_overview.mdx#in-band-migrations). 
+Information on Out of bound migrations can also be found there.
+
+
+Migrator with its relevant artifacts in the sourcegraph/sourcegraph repo can be viewed as an orchestrator with two special functions --
+
+1.  Migrator constructs migration plans given version ranges and a table
+    of migrations which have been successfully applied (each schema has
+    a table to track applied migrations within that schema). This logic
+    is supported by a variety of files generated during releases and
+    depends on the parent/child metadata of migrations generated via the
+    sg migration tool.
+
+2.  Migrator manages out of band migrations. These are data migrations
+    that must be run within specific schema boundaries. Running OOB
+    migrations at/after the deprecated version is unsupported. Migrator
+    ensures that the necessary OOB migrations are run at [stopping
+    points](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/oobmigration/upgrade.go?L32-42)
+    in a multiversion upgrade -- learn more
+    [here](https://about.sourcegraph.com/blog/introducing-migrator-service).
+
+### CLI design
+
+Migrator is designed with a CLI tool interface -- taking various commands to alter
+the state of the database and apply necessary schema changes. This
+design was initially implemented as a tool for TS team members to assist
+in multiversion upgrades and because it could easily be included over
+multiple deployment methods as a containerized process run separately
+from the frontend during startup ([which originally caused
+issues](https://about.sourcegraph.com/blog/introducing-migrator-service)).
+This replaced earlier go/migrate based strategies which ran in the
+frontend on startup. While the migrator can operate as a CLI tool, it's
+containerized as if it was another application which allows it to be run
+as an initContainer (in Kubernetes deployments) or as a dependent
+service (for Docker Compose deployments) to ensure that the necessary
+migrations are applied before startup of the application proper.
+Check out [RFC 469](https://docs.google.com/document/d/1_wqvNbEGhMzu45CHZNrLQYys7gl0AtVpLluGp4cbQKk/edit#heading=h.ddeuyk4t99yx) to learn more.
+
+### Important Commands
+
+The most important migrator commands are `up` and `upgrade`, with a notable mention to `drift`:
+
+-   **Up** : The default command of `migrator`, `up` ensures that
+    for a given version of the migrator, every migration defined at 
+    that build has been successfully applied in the connected database, 
+    this is specifically important to ensure *patch version migrations are run*. 
+    ***This must be run with the target version migrator image/build**.* 
```

Comment:
Maybe a "for example..." here showing what is meant.

Reply:
Gave an example, this is a tricky part of migrator 😓